### DISCUSSION TITLE
代码覆盖率：Midwayjs做的服务，非单元测试时如何得到覆盖率报告？使用nyc命令基本无法得到数据

### DISCUSSION THREAD

**Discussion by googoo625** [midwayjs/midway#4049] [Q&A]
### Describe the problem（描述问题）

我希望对Midwayjs实现的服务做系统测试（不是单元测试），然后生成覆盖率报告。

1）用 midway-bin cov --ts 命令启动程序，会直接报告没有tests而退出。

2）用nyc，启动命令：nyc --reporter=html --reporter=text npm run dev，可以正常启动。
测试后，用命令生成测试报告：npx nyc report --reporter=html --reporter=text --report-dir=./cov
可以生成报告。但是报告中的数字基本都是0。
如下是其中一个Controller/Service的截图（其他也都这样）。

![17253475601426](https://github.com/user-attachments/assets/ab83915e-80c6-4062-8924-254a454a08f4)

猜测可能是在midwayjs的Service注入模式下，nyc无法跟踪到代码调用。
还是希望直接用midwayjs的覆盖率工具。

请教这种情况下该如何配置？

### Midway Versions（Midway 版本）

3.16.2

**Comment by czy88840616** [comment]
jest --coverage ? 不过这是测试，我没见过 dev 还要算覆盖率的这个玩法..

**Reply by googoo625** [reply]
不行吧？这不是单元测试，没有test文件。
我只是用dev试一下，后面还是会用npm run start这种的。
是希望正常启动服务，然后在外面通过其他工具调用接口进行测试，结束时生成覆盖率报告。
nyc命令一般是可以做到的，但现在得到的数字异常，我觉得是因为Service是注入的，nyc没能监测到。

**Reply by czy88840616** [reply]
注入本质上就是api调用，跟覆盖率没有任何关系（毕竟单测走的也是注入）。覆盖率只跟覆盖率工具的插桩和计算有关。

**Reply by googoo625** [reply]
那我再调调nyc配置，可能有不对。不过咨询AI 确实提示“确保你的代码在运行时能够正确加载到 nyc 的覆盖范围之内，有时候依赖注入或动态引入的代码可能无法被 nyc 捕捉到。”
本来以为Midwayjs框架在这方面有改造，现在看来没有，那我就再试试nyc啦。

**Comment by waitingsong** [comment]
覆盖率是拿写好的 suit/case 去跑。lz 相当于起了一个服务（不管是 prod 还是 dev），然后去跑覆盖？ 好像没这么搞的吧～

**Reply by googoo625** [reply]
单元测试覆盖率之外，还想要系统测试覆盖率，算正常吧。在Java项目做过，所以现在想做NodeJS项目的。

**Reply by czy88840616** [reply]
就算是集成测试。。也是有 case 在那里，直接跑的。。而不是起动一个项目，再从外部触发

**Comment by waitingsong** [comment]
单测也是能覆盖到框架/系统的嘛，不行就新写个

**Reply by czy88840616** [reply]
node_modules的一般都忽略的。

**Reply by googoo625** [reply]
使用场景是这样的：
测试人员写一批自动化测试用例，在凌晨自动执行，执行完后进行覆盖率统计。

首先测试人员是用工具（如接口测试工具）实现的自动化用例，
其次系统在凌晨会重新部署（保证只有自动化测试影响了代码的覆盖率），
最后希望检查测试人员的自动化用例的覆盖程度。

**Comment by waitingsong** [comment]
```
首先测试人员是用工具（如接口测试工具）实现的自动化用例，
```
自动化**接口**测试工具并不能检测出测试覆盖率。 
代码覆盖率是需要用框架测试框架启动的服务上跑测试的。
